Add a `ListRerank` document compressor

denver1117 opened the pull request:
- **Description:** This PR adds a new document compressor called `ListRerank`. It's derived from `BaseDocumentCompressor`. It's a near exact implementation of introduced by this paper: [Zero-Shot Listwise Document Reranking with a Large Language Model](https://arxiv.org/pdf/2305.02156.pdf) which it finds to outperform pointwise reranking, which is somewhat implemented in LangChain as [LLMChainFilter](https://github.com/langchain-ai/langchain/blob/master/libs/langchain/langchain/retrievers/document_compressors/chain_filter.py). 
- **Issue:** None
- **Dependencies:** None
- **Tag maintainer:** @hwchase17 @izzymsft
- **Twitter handle:** @HarrisEMitchell

Notes:
1. I didn't add anything to `docs`. I wasn't exactly sure which patterns to follow as [cohere reranker is under Retrievers](https://python.langchain.com/docs/integrations/retrievers/cohere-reranker) with other external document retrieval integrations, but other contextual compression is [here](https://python.langchain.com/docs/modules/data_connection/retrievers/contextual_compression/). Happy to contribute to either with some direction.
2. I followed syntax, docstrings, implementation patterns, etc. as well as I could looking at nearby modules. One thing I didn't do was put the default prompt in a separate `.py` file like [Chain Filter](https://github.com/langchain-ai/langchain/blob/master/libs/langchain/langchain/retrievers/document_compressors/chain_filter_prompt.py) and [Chain Extract](https://github.com/langchain-ai/langchain/blob/master/libs/langchain/langchain/retrievers/document_compressors/chain_extract_prompt.py). Happy to follow that pattern if it would be preferred.

denver1117 committed:
add list reranker

denver1117 commented on libs/langchain/langchain/retrievers/document_compressors/list_rerank.py:
```diff
@@ -0,0 +1,99 @@
+            type="array[dict]",
+        )
+    ]
+    output_parser = StructuredOutputParser.from_response_schemas(response_schemas)
```
I experimented with a Pydantic parser that defines the full nested structure explicitly and saw notably more output parsing errors. Expressing the `array[dict]` type as an implicit nested type within a single `ResponseSchema` `type` argument was much more successful.

denver1117 committed:
add tests

hwchase17 reviewed (approved):
this seems really cool! Would be extra good to add a notebook in documentation as an example for this

denver1117 committed:
add example to documentation

denver1117 committed:
fix linting errors in tests

denver1117 committed:
sort imports

denver1117 committed:
fix typo in docs

denver1117 committed:
updated docs

denver1117 commented:
> this seems really cool! Would be extra good to add a notebook in documentation as an example for this

Thanks @hwchase17. I fixed the lint issues and added to the documentation. This is ready to go from my perspective.

hwchase17 committed:
cr

denver1117 committed:
fix arg order and callable type

denver1117 committed:
Merge branch 'feature/list-rerank-compressor' of github.com:denver1117/langchain into feature/list-rerank-compressor

denver1117 commented:
I fixed the order of arguments (need arg first, not kwarg) and fixed the bad type in the Callable arg as noted by the failed check.

baskaryan committed:
fmt

baskaryan committed:
fmt

baskaryan committed:
fmt

baskaryan committed:
poetry

baskaryan committed:
poetry

hwchase17 commented:
needs a re-review and clean up, but i generally like it

ccurme committed:
merge

ccurme committed:
move test

ccurme committed:
undo change to lock file

ccurme committed:
format

ccurme committed:
expand docstring

ccurme committed:
typing

ccurme committed:
remove unused typeddict

ccurme committed:
format

ccurme committed:
export LLMListwiseRerank from module

ccurme committed:
update docs

ccurme committed:
fix link typo
